src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs

```diff
@@ -4329,6 +4329,740 @@ static void Main()
+        }
+
+        [Fact]
+        public void LambdaReturnType_03()
+        {
+            var source =
+@"using System;
+class Program
+{
+    static void F<T, U>() where U : T
+    {
+        Func<T> f1;
+        Func<U> f2;
+        f1 = T () => default;
+        f2 = T () => default;
+        f1 = U () => default;
+        f2 = U () => default;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (9,14): error CS8918: Cannot convert lambda expression to type 'Func<U>' because the return type does not match the delegate return type
+                //         f2 = T () => default;
+                Diagnostic(ErrorCode.ERR_CantConvAnonMethReturnType, "T () => default").WithArguments("lambda expression", "System.Func<U>").WithLocation(9, 14),
+                // (10,14): error CS8918: Cannot convert lambda expression to type 'Func<T>' because the return type does not match the delegate return type
+                //         f1 = U () => default;
+                Diagnostic(ErrorCode.ERR_CantConvAnonMethReturnType, "U () => default").WithArguments("lambda expression", "System.Func<T>").WithLocation(10, 14));
+        }
+
+        [Fact]
+        public void LambdaReturnType_04()
+        {
+            var source =
+@"using System;
+using System.Linq.Expressions;
+class Program
+{
+    static void F<T, U>()
+    {
+        Expression<Func<T>> e1;
+        Expression<Func<U>> e2;
+        e1 = T () => default;
+        e2 = T () => default;
+        e1 = U () => default;
+        e2 = U () => default;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (10,14): error CS8918: Cannot convert lambda expression to type 'Expression<Func<U>>' because the return type does not match the delegate return type
+                //         e2 = T () => default;
+                Diagnostic(ErrorCode.ERR_CantConvAnonMethReturnType, "T () => default").WithArguments("lambda expression", "System.Linq.Expressions.Expression<System.Func<U>>").WithLocation(10, 14),
+                // (11,14): error CS8918: Cannot convert lambda expression to type 'Expression<Func<T>>' because the return type does not match the delegate return type
+                //         e1 = U () => default;
+                Diagnostic(ErrorCode.ERR_CantConvAnonMethReturnType, "U () => default").WithArguments("lambda expression", "System.Linq.Expressions.Expression<System.Func<T>>").WithLocation(11, 14));
+        }
+
+        [Fact]
+        public void LambdaReturnType_05()
+        {
+            var source =
+@"#nullable enable
+using System;
+class Program
+{
+    static void Main()
+    {
+        Func<dynamic> f1 = object () => default!;
+        Func<(int, int)> f2 = (int X, int Y) () => default;
```

Comment:
>(int X, int Y) () => default [](http://example.com/codeflow?start=30&length=28)

Conceptually, this feels similar to an override changing the names of a tuple. Do we want to consider a warning/error?

Reply:
We don't report diagnostics for simple variable assignments or for method group conversions and the lambda cases seem similar to those.